sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/FilterEstimation.scala

```diff
@@ -0,0 +1,555 @@
+    val mutableAttributeStats: mutable.Map[Attribute, ColumnStat] =
+      mutableColStats.map(kv => expridToAttrMap(kv._1) -> kv._2)
+    val newColStats = AttributeMap(mutableAttributeStats.toSeq)
+
+    val filteredRowCountValue: BigInt =
+      EstimationUtils.ceil(BigDecimal(stats.rowCount.get) * percent)
+    val avgRowSize = BigDecimal(EstimationUtils.getRowSize(plan.output, newColStats))
+    val filteredSizeInBytes: BigInt =
+      EstimationUtils.ceil(BigDecimal(filteredRowCountValue) * avgRowSize)
+
+    Some(stats.copy(sizeInBytes = filteredSizeInBytes, rowCount = Some(filteredRowCountValue),
+      attributeStats = newColStats))
+  }
+
+  /**
+   * Returns a percentage of rows meeting a compound condition in Filter node.
+   * A compound condition is depomposed into multiple single conditions linked with AND, OR, NOT.
+   * For logical AND conditions, we need to update stats after a condition estimation
+   * so that the stats will be more accurate for subsequent estimation.  This is needed for
+   * range condition such as (c > 40 AND c <= 50)
+   * For logical OR conditions, we do not update stats after a condition estimation.
+   *
+   * @param plan the Filter LogicalPlan node
+   * @param condition the compound logical expression
+   * @param update a boolean flag to specify if we need to update ColumnStat of a column
+   *               for subsequent conditions
+   * @return a doube value to show the percentage of rows meeting a given condition
+   */
+  def calculateConditions(
+      plan: Filter,
+      condition: Expression,
+      update: Boolean = true)
+    : Double = {
+
+    condition match {
+      case And(cond1, cond2) =>
+        val p1 = calculateConditions(plan, cond1, update)
+        val p2 = calculateConditions(plan, cond2, update)
+        p1 * p2
+
+      case Or(cond1, cond2) =>
+        val p1 = calculateConditions(plan, cond1, update = false)
+        val p2 = calculateConditions(plan, cond2, update = false)
+        math.min(1.0, p1 + p2 - (p1 * p2))
+
+      case Not(cond) => calculateSingleCondition(plan, cond, isNot = true, update = false)
+      case _ => calculateSingleCondition(plan, condition, isNot = false, update)
+    }
+  }
+
+  /**
+   * Returns a percentage of rows meeting a single condition in Filter node.
+   * Currently we only support binary predicates where one side is a column,
+   * and the other is a literal.
+   *
+   * @param plan the Filter LogicalPlan node
+   * @param condition a single logical expression
+   * @param isNot set to true for "IS NULL" condition.  set to false for "IS NOT NULL" condition
+   * @param update a boolean flag to specify if we need to update ColumnStat of a column
+   *               for subsequent conditions
+   * @return a doube value to show the percentage of rows meeting a given condition
+   */
+  def calculateSingleCondition(
+      plan: Filter,
+      condition: Expression,
+      isNot: Boolean,

```

Comment:
rather than passing isNot in here, I'd remove the argument and have this function return an Option[Double].

If it is an unsupported one, just return None.

Then in the few lines above, we can handle Not case directly there.